### DISCUSSION TITLE
How can Detox reproduce tap gestures and scrolling like a human hand?

### DISCUSSION THREAD

**Discussion by Wayne-Kim** [wix/Detox#4473] [Q&A]
How can Detox reproduce tap, gestures and scrolling like a human hand?

I wonder how this is possible internally.

There is so much code that I really don't know where to start.

Is there anyone who can give me some hints?

**Comment by portyu9** [comment]
Hi @Wayne-Kim 

Detox does not record and replay a literal human finger path.

At a high level, your JavaScript test expresses the action:

```js
await element(by.id('button')).tap()
await element(by.id('list')).swipe('up')
```

Detox sends that action to its native runtime, resolves the matched native view, and performs the corresponding platform UI action.

The other important part is synchronization: Detox waits for the application to become idle before continuing, instead of simply injecting coordinates followed by arbitrary sleeps.

Conceptually:

```text
JS test
  -> matcher/action
  -> Detox runtime communication
  -> native view resolution
  -> native tap/scroll/swipe
  -> synchronization
  -> next test action
```

For gestures, APIs such as `swipe()` also let you control direction, speed, amount, and starting position.

One important difference from older Detox descriptions is that current iOS Detox should not be thought of simply as "EarlGrey underneath everything"; modern Detox has its own native matcher/action infrastructure.

Please mark this answer as accepted if it helped, thank you.
